### PR TITLE
feat: favorites for appflowy

### LINKED ISSUE #443: [FR] Add to Favorites
**1~3 main use cases of the proposed feature**
Favoriting a page allows users to quickly access it without searching it or clicking through nested pages

**what types of users can benefit from using your proposed feature**
Users who have go-to pages that are frequently visited

**Additional context**
Key features:  
- Favorite a page: navigate to a page, then click a button at the top right of the page. 
- Favorites: a section in the left sidebar. It will appear when you favorite your first page. Here you can find pinned pages.
- Remove from Favorites: an option for the user to unpin a page

Scope: well-defined (see Key features) native application development

Difficulty level: Medium
Skills required: Flutter, Rust

### PR DESCRIPTION
**Pull request by squidrye**
### Feature Preview
Edit: Here is a brief demo of what I have so far

https://github.com/AppFlowy-IO/AppFlowy/assets/84044317/36e41dd9-ffba-483d-b9ee-8822aeb0ac07

---

#### PR Checklist

- [x] My code adheres to the [AppFlowy Style Guide](https://appflowy.gitbook.io/docs/essential-documentation/contribute-to-appflowy/software-contributions/submitting-code/style-guides)
- [x] I've listed at least one issue that this PR fixes in the description above.
- [x] I've added a test(s) to validate changes in this PR, or this PR only contains semantic changes.
- [ ] All existing tests are passing.

fixes #443 
depends on https://github.com/AppFlowy-IO/AppFlowy-Collab/pull/74

### CODE REVIEW

File: frontend/appflowy_flutter/lib/workspace/application/favorite/favorite_bloc.dart [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  const factory FavoriteEvent.didReceiveFavorite(List<FavoritesPB> favorite) =
+      ReceiveFavorites;
+  const factory FavoriteEvent.toggle(String viewId) = ToggleFavorite;
+}
+
+@freezed
+class FavoriteState with _$FavoriteState {
+  const factory FavoriteState({
+    required List<FavoritesPB> objects,
+    required Either<Unit, FlowyError> successOrFailure,
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/workspace/application/favorite/favorite_bloc.dart:82]: The right side should be the success, left side the failure. By common conventions.

File: frontend/appflowy_flutter/lib/workspace/application/favorite/favorite_listener.dart [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import 'package:appflowy_backend/protobuf/flowy-folder2/favorite.pb.dart';
+import 'package:appflowy_backend/protobuf/flowy-folder2/notification.pb.dart';
+import 'package:appflowy_backend/protobuf/flowy-notification/subject.pb.dart';
+import 'package:appflowy_backend/rust_stream.dart';
+import 'package:dartz/dartz.dart';
+import 'package:flutter/foundation.dart';
+
+class FavoriteListener {
+  StreamSubscription<SubscribeObject>? _streamSubscription;
+  void Function(Either<List<FavoritesPB>, FlowyError> _favoritesOrFailed)?
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/workspace/application/favorite/favorite_listener.dart:14]: Same here

File: frontend/appflowy_flutter/lib/workspace/presentation/home/menu/app/section/item.dart [hunk 3]
```diff
@@ -157,6 +162,8 @@ extension ViewDisclosureExtension on ViewDisclosureAction {
         return const FlowySvg(name: 'editor/delete');
       case ViewDisclosureAction.duplicate:
         return const FlowySvg(name: 'editor/copy');
+      case ViewDisclosureAction.favorite:
+        return const FlowySvg(name: 'editor/copy');
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/workspace/presentation/home/menu/app/section/item.dart:166]: We need to import the correct svg if it is not present.

File: frontend/appflowy_flutter/lib/workspace/presentation/home/menu/menu.dart [hunk 4]
```diff
@@ -104,6 +107,35 @@ class HomeMenu extends StatelessWidget {
...
+            visible: true,
+            child: ExpandableTheme(
+              data: ExpandableThemeData(
+                useInkWell: true,
+                animationDuration: Durations.medium,
+              ),
+              child: ScrollConfiguration(
+                behavior: const ScrollBehavior().copyWith(scrollbars: false),
+                child: Column(
+                  children: [...state.objects.map((e) => Text(e.id)).toList()],
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/workspace/presentation/home/menu/menu.dart:129]: You can remove the spread operator and the brackets `children: state.objects.map().toList(),`

File: frontend/appflowy_flutter/lib/workspace/presentation/home/menu/app/favorite_header.dart [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  @override
+  Widget build(BuildContext context) {
+    return SizedBox(
+      height: MenuAppSizes.headerHeight,
+      child: Row(
+        mainAxisAlignment: MainAxisAlignment.start,
+        crossAxisAlignment: CrossAxisAlignment.center,
+        children: [
+          _renderExpandedIcon(context),
+          const Text("Favorites"),
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/workspace/presentation/home/menu/app/favorite_header.dart:33]: Localization

File: frontend/appflowy_flutter/lib/workspace/presentation/home/menu/app/favorite_header.dart [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+            required: true,
+          )?.toggle();
+        },
+        child: ExpandableIcon(
+          theme: ExpandableThemeData(
+            expandIcon: FlowyIconData.drop_down_show,
+            collapseIcon: FlowyIconData.drop_down_hide,
+            iconColor: Theme.of(context).colorScheme.tertiary,
+            iconSize: MenuAppSizes.iconSize,
+            iconPadding: const EdgeInsets.fromLTRB(0, 0, 10, 0),
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/workspace/presentation/home/menu/app/favorite_header.dart:57]:
```suggestion
            iconPadding: const EdgeInsets.only(right: 10),
```

File: frontend/appflowy_flutter/lib/workspace/presentation/home/menu/menu.dart [hunk 7]
```diff
@@ -104,6 +108,56 @@ class HomeMenu extends StatelessWidget {
...
   }
 
+  Widget _renderFavorites(BuildContext context) {
+    return BlocProvider(
+      create: (context) =>
+          getIt<FavoriteBloc>()..add(const FavoriteEvent.initial()),
+      child: BlocConsumer<FavoriteBloc, FavoriteState>(
+        listener: (context, state) {
+          Log.warn("Favorite state changed $state");
+        },
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/workspace/presentation/home/menu/menu.dart:118]: Once ready for review/merge, if the listener isn't being used, then prefer a BlocBuilder over Consumer. Maybe add a TODO for yourself to not forget :)

File: frontend/appflowy_flutter/lib/workspace/presentation/home/menu/menu.dart [hunk 8]
```diff
@@ -104,6 +108,56 @@ class HomeMenu extends StatelessWidget {
...
+                                key: ValueKey(e.id),
+                                isSelected: false,
+                                onSelected: (item) {},
+                                view: e,
+                              ),
+                            )
+                            .toList(),
+                      ),
+                    ),
+                    collapsed: const SizedBox(),
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/workspace/presentation/home/menu/menu.dart:152]:
```suggestion
                    collapsed: const SizedBox.shrink(),
```

File: frontend/appflowy_flutter/lib/workspace/presentation/home/menu/menu.dart [hunk 9]
```diff
@@ -104,6 +108,56 @@ class HomeMenu extends StatelessWidget {
...
+                  ),
+                )
+              : const Offstage();
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/workspace/presentation/home/menu/menu.dart:155]: Why `Offstage` here rather than `const SizedBox.shrink()`, Offstage is worse for performance.

File: frontend/appflowy_flutter/lib/workspace/presentation/home/menu/menu.dart [hunk 10]
```diff
@@ -121,17 +175,24 @@ class HomeMenu extends StatelessWidget {
...
-                      EdgeInsets.only(bottom: 20.0 - MenuAppSizes.appVPadding),
-                  child: MenuUser(user),
+                header: Column(
+                  children: [
+                    Padding(
+                      padding: EdgeInsets.only(
+                        bottom: 20.0 - MenuAppSizes.appVPadding,
+                      ),
+                      child: MenuUser(user),
+                    ),
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/workspace/presentation/home/menu/menu.dart:185]: I know you didn't add this, but this seems quite weird. `20.0` is a magic number, and subtracting a constant doesn't make it less magic... 
```suggestion
                    Padding(
                      padding: EdgeInsets.only(bottom: 6),
                      child: MenuUser(user),
                    ),
```

File: frontend/appflowy_flutter/lib/workspace/presentation/home/menu/app/favorite_header.dart (hunk 5 above)
Comment (squidrye) [reply frontend/appflowy_flutter/lib/workspace/presentation/home/menu/app/favorite_header.dart:33]: Oh I missed this, I'll see to it.

**Comment by squidrye** [comment]
I'll drop in a bit of update here, for most part feature is built, list of tasks that remain here can be summarised as following

1. Remove redundant notifiers from rust lib
2. Handle edge case for when views are deleted.
3. Update the flutter code for some maintenance tasks, I'll take care of all the suggestions there, thanks for the review! @Xazin 
4. Tests

This should be ready for release by weekend.

File: frontend/rust-lib/flowy-folder2/src/entities/view.rs [hunk 11]
```diff
@@ -87,6 +91,7 @@ pub fn view_pb_with_child_views(view: Arc<View>, child_views: Vec<Arc<View>>) ->
     layout: view.layout.clone().into(),
     icon_url: view.icon_url.clone(),
     cover_url: view.cover_url.clone(),
+    is_favorite: false,
```
Comment (Xazin) [inline frontend/rust-lib/flowy-folder2/src/entities/view.rs:94]:
```suggestion
    is_favorite: view.is_favorite,
```

File: frontend/rust-lib/flowy-folder2/src/entities/view.rs [hunk 12]
```diff
@@ -303,10 +308,13 @@ pub struct UpdateViewPayloadPB {
   pub layout: Option<ViewLayoutPB>,
 
   #[pb(index = 6, one_of)]
-  pub icon_url: Option<String>,
+  pub cover_url: Option<String>,
 
   #[pb(index = 7, one_of)]
-  pub cover_url: Option<String>,
+  pub icon_url: Option<String>,
```
Comment (Xazin) [inline frontend/rust-lib/flowy-folder2/src/entities/view.rs:314]: Did you accidentally change the order of these two?

File: frontend/rust-lib/flowy-folder2/src/event_handler.rs [hunk 13]
```diff
@@ -1,11 +1,11 @@
...
 
-use flowy_error::FlowyError;
-use lib_dispatch::prelude::{data_result_ok, AFPluginData, AFPluginState, DataResult};
-
 use crate::entities::*;
 use crate::manager::FolderManager;
 use crate::share::ImportParams;
+use flowy_error::FlowyError;
+use lib_dispatch::prelude::{data_result_ok, AFPluginData, AFPluginState, DataResult};
+// use tracing::instrument;
```
Comment (Xazin) [inline frontend/rust-lib/flowy-folder2/src/event_handler.rs:8]: Remove commented out code

File: frontend/rust-lib/flowy-folder2/src/event_handler.rs [hunk 14]
```diff
@@ -132,7 +132,16 @@ pub(crate) async fn delete_view_handler(
...
   Ok(())
 }
-
+pub(crate) async fn toggle_favorites_handler(
+  data: AFPluginData<RepeatedViewIdPB>,
+  folder: AFPluginState<Arc<FolderManager>>,
+) -> Result<(), FlowyError> {
+  let params: RepeatedViewIdPB = data.into_inner();
+  for view_id in &params.items {
+    let _ = folder.toggle_favorites(view_id).await;
```
Comment (Xazin) [inline frontend/rust-lib/flowy-folder2/src/event_handler.rs:141]:
```suggestion
    folder.toggle_favorites(view_id).await;
```

File: frontend/rust-lib/flowy-folder2/src/entities/view.rs (hunk 11 above)
Comment (squidrye) [reply frontend/rust-lib/flowy-folder2/src/entities/view.rs]: This helped me figure out a major bug, thanks!

**Comment by squidrye** [comment]
Hey, @Xazin added some tests on flutter side, I'll add tests for rust in upcoming days, that should mark the base feature complete, let me know if refactor is needed for any part of my approach.

File: frontend/appflowy_flutter/lib/workspace/presentation/home/menu/menu.dart [hunk 15]
```diff
@@ -105,6 +112,50 @@ class HomeMenu extends StatelessWidget {
...
+                              isSelected: false,
+                              onSelected: (view) => getIt<MenuSharedState>()
+                                  .latestOpenView = view,
+                              view: e,
+                            ),
+                          )
+                          .toList(),
+                    ),
+                  ),
+                  collapsed: const SizedBox(),
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/workspace/presentation/home/menu/menu.dart:151]:
```suggestion
                  collapsed: const SizedBox.shrink(),
```

**Comment by LucasXu0** [comment]
@squidrye FYI, the sidebar has been refactored in this PR. https://github.com/AppFlowy-IO/AppFlowy/pull/3083

**Comment by squidrye** [comment]
oh okay, I'll rebase on top of your branch to ensure favorites work fine with nestings @LucasXu0

**Comment by LucasXu0** [comment]
Please rebase the latest main branch after this [PR](https://github.com/AppFlowy-IO/AppFlowy/pull/3083) merged.

Here're some suggestions:

- [ ] Remove the `_renderFavorites` function and replace it with a new widget to build the favorite list, and add the new widget inside `SidebarFolder`.
- [ ] Make sure that drag and drop works fine.

**Comment by LucasXu0** [comment]
@squidrye The pull request for the draggable files list has been merged. You can rebase or merge the main branch to continue finishing this feature.

FYI, the menu/section/item are deprecated. Please use sidebar/view_item instead.

**Comment by squidrye** [comment]
Got it, I'll refactor it as soon as I get some time.

File: frontend/rust-lib/flowy-folder2/src/notification.rs [hunk 16]
```diff
@@ -57,6 +60,8 @@ impl std::convert::From<i32> for FolderNotification {
       15 => FolderNotification::DidUpdateTrash,
       16 => FolderNotification::DidUpdateFolderSnapshotState,
       17 => FolderNotification::DidUpdateFolderSyncUpdate,
+      36 => FolderNotification::DidFavoriteView,
+      37 => FolderNotification::DidUnFavoriteView,
```
Comment (Xazin) [inline frontend/rust-lib/flowy-folder2/src/notification.rs:64]:
```suggestion
      37 => FolderNotification::DidUnfavoriteView,
```

File: frontend/rust-lib/flowy-folder2/src/manager.rs [hunk 17]
```diff
@@ -634,6 +656,60 @@ impl FolderManager {
...
+  }
+
+  #[tracing::instrument(level = "trace", skip(self))]
+  pub(crate) async fn get_all_favorites(&self) -> Vec<FavoritesInfo> {
+    self.with_folder(vec![], |folder| {
+      let trash_ids = folder
+        .get_all_trash()
+        .into_iter()
+        .map(|trash| trash.id)
+        .collect::<Vec<String>>();
```
Comment (Xazin) [inline frontend/rust-lib/flowy-folder2/src/manager.rs:705]: Fix

File: frontend/rust-lib/flowy-folder2/src/manager.rs [hunk 18]
```diff
@@ -634,6 +656,60 @@ impl FolderManager {
...
+            .payload(RepeatedViewPB {
+              items: vec![view_pb_without_child_views(new_view)],
+            })
+            .send();
+        } else {
+          send_notification("favorite", FolderNotification::DidUnFavoriteView)
+            .payload(RepeatedViewPB {
+              items: vec![view_pb_without_child_views(new_view)],
+            })
+            .send();
```
Comment (Xazin) [inline frontend/rust-lib/flowy-folder2/src/manager.rs:685]: Can be reduced

File: frontend/rust-lib/flowy-folder2/src/manager.rs [hunk 19]
```diff
@@ -442,21 +442,42 @@ impl FolderManager {
...
+              .iter()
+              .map(|v| v.id.clone())
+              .collect(),
+          );
+          send_notification("favorite", FolderNotification::DidUnFavoriteView)
+            .payload(RepeatedViewPB {
+              items: favorite_descendant_views,
+            })
+            .send();
+        }
```
Comment (Xazin) [inline frontend/rust-lib/flowy-folder2/src/manager.rs:471]: This could be it's own method, as it clutters the `move_view_to_trash`

File: frontend/appflowy_flutter/lib/workspace/application/favorite/favorite_bloc.dart [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+          },
+          didUnfavorite: (e) {
+            final list = List<ViewPB>.from(state.objects);
+            for (final item in e.favorite.items) {
+              list.removeWhere((element) => element.id == item.id);
+            }
+            emit(
+              state.copyWith(objects: list),
+            );
+          },
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/workspace/application/favorite/favorite_bloc.dart:53]:
```suggestion
          didUnfavorite: (e) {
            final views = [...state.views]..removeWhere(
                (view) => e.favorite.items.any((item) => item.id == view.id),
              );
            emit(
              state.copyWith(views: views),
            );
          },
```

File: frontend/appflowy_flutter/lib/workspace/presentation/home/menu/sidebar/sidebar.dart [hunk 21]
```diff
@@ -33,11 +35,19 @@ class HomeSideBar extends StatelessWidget {
...
+      providers: [
+        BlocProvider(
+          create: (_) => MenuBloc(
+            user: user,
+            workspace: workspaceSetting.workspace,
+          )..add(const MenuEvent.initial()),
+        ),
+        BlocProvider(
+          create: (_) =>
+              getIt<FavoriteBloc>()..add(const FavoriteEvent.initial()),
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/workspace/presentation/home/menu/sidebar/sidebar.dart:48]:
```suggestion
          create: (_) => FavoriteBloc()..add(const FavoriteEvent.initial()),
```

No need to use GetIt here.

File: frontend/appflowy_flutter/lib/workspace/presentation/home/menu/sidebar/sidebar.dart [hunk 22]
```diff
@@ -69,6 +79,9 @@ class HomeSideBar extends StatelessWidget {
             SidebarUser(user: user),
             // Favorite, Not supported yet
             const VSpace(20),
+            const SingleChildScrollView(
+              child: SidebarFavorite(),
+            ),
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/workspace/presentation/home/menu/sidebar/sidebar.dart:84]: Move it inside the sidebar folder as an independent folder.

File: frontend/appflowy_flutter/lib/workspace/presentation/home/menu/sidebar/sidebar_favorite.dart [hunk 23]
```diff
@@ -0,0 +1,98 @@
+import 'package:appflowy/generated/locale_keys.g.dart';
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/workspace/presentation/home/menu/sidebar/sidebar_favorite.dart:1]: remove this file and create a new one inside the sidebar/folder.

```dart
  Widget build(BuildContext context) {
    return Column(
      mainAxisAlignment: MainAxisAlignment.start,
      children: [
        // favorite
        const FavoriteFolder(),
        // personal
        PersonalFolder(views: views),
      ],
    );
  }
```

File: frontend/rust-lib/flowy-folder2/src/entities/view.rs [hunk 24]
```diff
@@ -345,14 +350,20 @@ impl TryInto<UpdateViewParams> for UpdateViewPayloadPB {
       Some(thumbnail) => Some(ViewThumbnail::parse(thumbnail)?.0),
     };
 
+    let cover_url = self.cover_url;
+
```
Comment (LucasXu0) [inline frontend/rust-lib/flowy-folder2/src/entities/view.rs:354]: remove this empty line.

File: frontend/rust-lib/flowy-folder2/src/event_handler.rs [hunk 25]
```diff
@@ -151,7 +150,17 @@ pub(crate) async fn delete_view_handler(
...
+  data: AFPluginData<RepeatedViewIdPB>,
+  folder: AFPluginState<Weak<FolderManager>>,
+) -> Result<(), FlowyError> {
+  let params: RepeatedViewIdPB = data.into_inner();
+  let folder = upgrade_folder(folder)?;
+  for view_id in &params.items {
+    let _ = folder.toggle_favorites(view_id).await;
+  }
+  Ok(())
+}
```
Comment (LucasXu0) [inline frontend/rust-lib/flowy-folder2/src/event_handler.rs:163]:
```suggestion

pub(crate) async fn toggle_favorites_handler(
  data: AFPluginData<RepeatedViewIdPB>,
  folder: AFPluginState<Weak<FolderManager>>,
) -> Result<(), FlowyError> {
  let params: RepeatedViewIdPB = data.into_inner();
  let folder = upgrade_folder(folder)?;
  for view_id in &params.items {
    let _ = folder.toggle_favorites(view_id).await;
  }
  Ok(())
}

```

File: frontend/appflowy_flutter/lib/workspace/application/favorite/favorite_listener.dart [hunk 26]
```diff
@@ -0,0 +1,66 @@
...
+              _favoriteUpdated!(right(unfavoriteView), false);
+            },
+            (error) => _favoriteUpdated!(left(error), false),
+          );
+        }
+        break;
+      default:
+        break;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (LucasXu0) [reply frontend/appflowy_flutter/lib/workspace/application/favorite/favorite_listener.dart:59]: don't commit, just for recording.

File: frontend/appflowy_flutter/lib/workspace/application/favorite/favorite_bloc.dart (hunk 20 above)
Comment (LucasXu0) [reply frontend/appflowy_flutter/lib/workspace/application/favorite/favorite_bloc.dart:53]: don't commit, just for recording.

File: frontend/appflowy_flutter/lib/workspace/presentation/home/menu/sidebar/sidebar.dart (hunk 21 above)
Comment (LucasXu0) [reply frontend/appflowy_flutter/lib/workspace/presentation/home/menu/sidebar/sidebar.dart:48]: don't commit, just for recording.

**Comment by LucasXu0** [comment]
https://github.com/AppFlowy-IO/AppFlowy/assets/11863087/b227c12e-75fa-42c0-ad74-67c00286cea5

For now, the selection of the same page will be synced, but the rename function doesn't work.

**Comment by LucasXu0** [comment]
FYI, I updated the collab repo and now it supports creating the view at the first index of the personal list instead of the last one.

**Comment by qinluhe** [comment]
Please update  `frontend/appflowy_tauri/src-tauri/Cargo.toml`
image

File: frontend/rust-lib/flowy-folder2/src/manager.rs (hunk 19 above)
Comment (squidrye) [reply frontend/rust-lib/flowy-folder2/src/manager.rs]: updated!

File: frontend/appflowy_flutter/lib/workspace/presentation/home/menu/view/view_action_type.dart [hunk 27]
```diff
@@ -38,12 +38,12 @@ extension ViewMoreActionTypeExtension on ViewMoreActionType {
     }
   }
 
-  Widget icon(Color iconColor, {ViewState? state}) {
+  Widget icon(Color iconColor, {bool? state}) {
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/workspace/presentation/home/menu/view/view_action_type.dart]:
```suggestion
  Widget icon(Color iconColor, {bool? isFavorited}) {
```

File: frontend/appflowy_flutter/lib/workspace/presentation/home/menu/view/view_action_type.dart [hunk 28]
```diff
@@ -38,12 +38,12 @@ extension ViewMoreActionTypeExtension on ViewMoreActionType {
...
     switch (this) {
       case ViewMoreActionType.delete:
         return const FlowySvg(name: 'editor/delete');
       case ViewMoreActionType.toggleFavorite:
-        if (state!.view.isFavorite) {
+        if (state!) {
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/workspace/presentation/home/menu/view/view_action_type.dart]:
```
        return const FlowySvg(name: isFavorited! ? 'home/unfavorite' : 'home/favorite');
```
or 
```
        return const FlowySvg(name: 'home/${isFavorited! ? 'unfavorite' : 'favorite'}');
```
or
```
        return const FlowySvg(name: 'home/${isFavorited! ? 'un' : ''}favorite');
```

Comment (LucasXu0) [reply frontend/appflowy_flutter/lib/workspace/presentation/home/menu/view/view_action_type.dart]: It's outdated; I have refactored this part.

File: frontend/appflowy_flutter/lib/workspace/presentation/home/menu/menu.dart [hunk 29]
```diff
@@ -105,6 +111,50 @@ class HomeMenu extends StatelessWidget {
     );
   }
 
+  Widget _renderFavorites(BuildContext context) {
+    return BlocBuilder<FavoriteBloc, FavoriteState>(
+      builder: (context, state) {
+        return state.views.isNotEmpty
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/workspace/presentation/home/menu/menu.dart:117]: I would prefer not to use the ternary operator on such a large widget. Up to preference though.
```
if (state.views.isEmpty) {
  return const SizedBox.shrink();
}

return ExpandableTheme(
```

Comment (squidrye) [reply frontend/appflowy_flutter/lib/workspace/presentation/home/menu/menu.dart:117]: Actually this code is deprecated, all the logic for favorites and views is now under sidebar/sidebar.dart and sidebar/folder/favorites_folder.dart.

I think removal of deprecated code will be handled by lucas after making sure it does not impact anything else
